File: tests/test_mi_with_explain.pl
% Test meta-interpreter with explain_goal pattern from cans.pl

fact(a).
rule(X) :- fact(X).

% Explanation for rule
explain_goal(rule(X), Msg) :-
    format(atom(Msg), "Checking rule for ~w", [X]).

% Meta-interpreter with explanation (from cans.pl)
mi(true, Expl, Expl) :- !.

mi((A,B), ExplIn, ExplOut) :- !,
    mi(A, ExplIn, ExplMid),
    mi(B, ExplMid, ExplOut).

mi(Goal, ExplIn, ExplOut) :-
    explain_goal(Goal, Msg), !,
    append(ExplIn, [Msg], ExplMid),
    (   clause(Goal, Body)
    ->  mi(Body, ExplMid, ExplOut)
    ;   call(Goal),
        ExplOut = ExplMid
    ).

mi(Goal, ExplIn, ExplOut) :-
    (   clause(Goal, Body)
    ->  mi(Body, ExplIn, ExplOut)
    ;   call(Goal),
        ExplOut = ExplIn
    ).

File: tests/test_mi_with_explain.py
from vibeprolog import PrologInterpreter

prolog = PrologInterpreter()
prolog.consult("tests/test_mi_with_explain.pl")

print("1. Test explain_goal(rule(X), Msg)...")
results = prolog.query("explain_goal(rule(X), Msg).")
print(f"   Results: {results[:1]}")

print("\n2. Test mi(rule(X), [], Expl)...")
results = prolog.query("mi(rule(X), [], Expl).")
print(f"   Results: {results[:2]}")

print("\n3. Test mi(rule(a), [], Expl)...")
results = prolog.query("mi(rule(a), [], Expl).")
print(f"   Results: {results}")

print("\n4. Test mi(fact(a), [], Expl)...")
results = prolog.query("mi(fact(a), [], Expl).")
print(f"   Results: {results}")
